### PR TITLE
Interlinks Autocomplete: Mark I

### PR DESCRIPTION
**Pull request by jleandroperez**
### Fix
In this PR we're implementing **InterlinkResultsController**. This component will be in charge of:

1. Filtering Notes that contain a specified Interlink keyword
2. Excluding a specified Entity ID (useful to exclude the "current" document, and thus, prevent loop links)

We're also adding capabilities to limit the number of results.

Please bear in mind this is part of a larger feature, splitting this into review friendly pieces.

Ref. #914
cc @eshurakov (Thanks in advance!!)

### Test
- [x] Verify the code makes sense
- [x] Verify the new unit tests are green!
- [x] Yay!

### Release
These changes do not require release notes.

### CODE REVIEW

File: Simplenote.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -88,7 +88,6 @@
 		46A3C99B17DFA81A002865AE /* Tag.m in Sources */ = {isa = PBXBuildFile; fileRef = 467D9C641788A54900785EF3 /* Tag.m */; };
 		46A3C99C17DFA81A002865AE /* NSString+Bullets.m in Sources */ = {isa = PBXBuildFile; fileRef = E2C5C73F17CAD2FD003B9795 /* NSString+Bullets.m */; };
 		46A3C99D17DFA81A002865AE /* SPAddCollaboratorsViewController.m in Sources */ = {isa = PBXBuildFile; fileRef = E25C39CF17A41F3400B2591A /* SPAddCollaboratorsViewController.m */; };
-		46A3C9A117DFA81A002865AE /* SPTagsListViewController.m in Sources */ = {isa = PBXBuildFile; fileRef = E277B5E8179EDAFB0095CD24 /* SPTagsListViewController.m */; };
```
Comment (jleandroperez) [inline Simplenote.xcodeproj/project.pbxproj]: I've just noticed Xcode did this. I'll take care of fixing merge conflicts (if they do arise!).

Thank you!

File: Simplenote/Classes/InterlinkResultsController.swift [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        resultsController.predicate = NSPredicate.predicateForNotes(deleted: false)
+        try? resultsController.performFetch()
+    }
+
+    /// Returns the collection of Notes filtered by the specified Keyword in their title, excluding a specific ObjectID
+    /// - Important: Returns `nil` when there are no results!
+    ///
+    func searchNotes(byTitleKeyword keyword: String, excluding excludedID: NSManagedObjectID?) -> [Note]? {
+        filter(notes: resultsController.fetchedObjects, byTitleKeyword: keyword, excluding: excludedID)
+    }
```
Comment (eshurakov) [inline Simplenote/Classes/InterlinkResultsController.swift:34]: Does it search with only one keyword? I though maybe it's nice to have a behaviour similar to the main search: split by words and search for all of them.

File: Simplenote/Classes/InterlinkResultsController.swift [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    ///     - It's easier and anyone can follow along!
+    ///
+    func filter(notes: [Note], byTitleKeyword keyword: String, excluding excludedID: NSManagedObjectID?) -> [Note]? {
+        let normalizedKeyword = keyword.folding(options: [.diacriticInsensitive, .caseInsensitive], locale: nil)
+        var output = [Note]()
+
+        for note in notes where note.objectID != excludedID {
+            note.ensurePreviewStringsAreAvailable()
+            guard let normalizedTitle = note.titlePreview?.folding(options: [.diacriticInsensitive, .caseInsensitive], locale: nil),
+                  normalizedTitle.contains(normalizedKeyword)
```
Comment (eshurakov) [inline Simplenote/Classes/InterlinkResultsController.swift]: I believe you can do it without folding keyword and title.
```
guard note.titlePreview.range(of: keyword, options: [.caseInsensitive, .diacriticInsensitive], range: note.titlePreview.startIndex..<note.titlePreview.endIndex, locale: Locale.current) != nil else {
    continue
}
```

**Review by eshurakov — APPROVED** [review]
Looks good! 👍

File: Simplenote/Classes/InterlinkResultsController.swift (hunk 2 above)
Comment (jleandroperez) [reply Simplenote/Classes/InterlinkResultsController.swift:34]: You're **SO** right!! updating shortly!! I can't believe I missed that. Thank you for pointing it out!!

File: Simplenote/Classes/InterlinkResultsController.swift (hunk 3 above)
Comment (jleandroperez) [reply Simplenote/Classes/InterlinkResultsController.swift]: MUCH better!! thanks for the tip!

**Comment by jleandroperez** [comment]
@eshurakov Mind taking another peek, just for safety?

Thanks in advance, loved the feedback!!

**Comment by jleandroperez** [comment]
### Note:
I've missed a huge detail. When searching for a document, if we search something that contains **any** of the words the user entered, the filtering is a bit shaky.

Why: Suppose your note's title contains `It's a long way to the top`.
When your document looks like `[it's a`, we'd be showing documents that contain both `it's` and `a` (in my case that translates to... way above 10 documents!).

Let's revert to the "single word" behavior (plus it matches macOS!). Thank you!

File: Simplenote/Classes/String+Simplenote.swift [hunk 4]
```diff
@@ -23,6 +23,12 @@ extension String {
 //
 extension String {
 
+    /// Returns the Range enclosing all of the receiver's contents
+    ///
+    var fullRange: Range<String.Index> {
+        self.startIndex ..< self.endIndex
```
Comment (eshurakov) [inline Simplenote/Classes/String+Simplenote.swift]: Do you need `self` here? 🤔

Comment (jleandroperez) [reply Simplenote/Classes/String+Simplenote.swift]: Thank you!!

**Comment by jleandroperez** [comment]
Merging this one. Thank you @eshurakov!!

**Pull request merged**